zenith_utils/src/pq_proto.rs

```diff
@@ -169,44 +164,53 @@ impl FeStartupMessage {
             bail!("invalid message length");
         }
 
-        let version = stream.read_u32::<BE>()?;
-        let kind = match version {
-            CANCEL_REQUEST_CODE => StartupRequestCode::Cancel,
-            NEGOTIATE_SSL_CODE => StartupRequestCode::NegotiateSsl,
-            NEGOTIATE_GSS_CODE => StartupRequestCode::NegotiateGss,
-            _ => StartupRequestCode::Normal,
-        };
+        let request_code = stream.read_u32::<BE>()?;
 
         // the rest of startup packet are params
         let params_len = len - 8;
         let mut params_bytes = vec![0u8; params_len];
         stream.read_exact(params_bytes.as_mut())?;
 
-        // Then null-terminated (String) pairs of param name / param value go.
-        let params_str = str::from_utf8(&params_bytes).unwrap();
-        let params = params_str.split('\0');
-        let mut params_hash: HashMap<String, String> = HashMap::new();
-        for pair in params.collect::<Vec<_>>().chunks_exact(2) {
-            let name = pair[0];
-            let value = pair[1];
-            if name == "options" {
-                // deprecated way of passing params as cmd line args
-                for cmdopt in value.split(' ') {
-                    let nameval: Vec<&str> = cmdopt.split('=').collect();
-                    if nameval.len() == 2 {
-                        params_hash.insert(nameval[0].to_string(), nameval[1].to_string());
+        // Parse params depending on request code
+        let message = match request_code {
+            CANCEL_REQUEST_CODE => {
+                let parse_i32_big_endian = |byte_slice: &[u8]| {
+                    let byte_array: [u8; 4] = byte_slice.try_into().unwrap();
+                    i32::from_be_bytes(byte_array)
+                };
+
+                ensure!(params_len == 8, "expected 8 bytes for CancelRequest params");
+                let backend_pid = parse_i32_big_endian(&params_bytes[0..4]);
+                let cancel_key = parse_i32_big_endian(&params_bytes[4..8]);
```

Comment:
Try using `byteorder`'s `read_u32::<BE>` on the buffer (might need to wrap with a `Cursor`?)